ReactiveCocoa/Swift/Flatten.swift

```diff
@@ -341,25 +341,52 @@ extension SignalProtocol where Value: SignalProducerProtocol, Error == Value.Err
 			return disposable
 		}
 	}
-
+	
 	fileprivate func observeConcat(_ observer: Observer<Value.Value, Error>, _ disposable: CompositeDisposable? = nil) -> Disposable? {
-		let state = ConcatState(observer: observer, disposable: disposable)
+		let state = Atomic(ConcatState<Value.Value, Error>())
+		
+		func startNextIfNeeded() {
+			while let producer = state.modify({ $0.dequeue() }) {
+				producer.startWithSignal { signal, inner in
+					let handle = disposable?.add(inner) ?? nil
+
+					signal.observe { event in
+						switch event {
+						case .completed, .interrupted:
+							handle?.remove()
+							
+							state.modify { $0.active = nil }
 
-		return self.observe { event in
+							if !state.value.isStarting {
```

Comment:
You can return `isStarting` from the action above to avoid acquiring the lock twice.